# Reason: Let modules exempt their control spells before native DR increment and immunity.
# Approved: ruling 2026-09-13.
--- a/src/server/game/Scripting/ScriptDefines/UnitScript.cpp
+++ b/src/server/game/Scripting/ScriptDefines/UnitScript.cpp
@@ -177,3 +177,11 @@ UnitScript::UnitScript(char const* name, bool addToScripts, std::vector<uint16>
 }
 
 template class AC_GAME_API ScriptRegistry<UnitScript>;
+
+bool ScriptMgr::IsSpellDiminishingExempt(Spell* cast, Unit* caster, Unit* target, SpellInfo const* info)
+{
+    for (auto const& script : ScriptRegistry<UnitScript>::EnabledHooks[UNITHOOK_IS_SPELL_DIMINISHING_EXEMPT])
+        if (script->IsSpellDiminishingExempt(cast, caster, target, info))
+            return true;
+    return false;
+}
--- a/src/server/game/Scripting/ScriptDefines/UnitScript.h
+++ b/src/server/game/Scripting/ScriptDefines/UnitScript.h
@@ -46,6 +46,7 @@ enum UnitHook
     UNITHOOK_ON_UNIT_SET_SHAPESHIFT_FORM,
     UNITHOOK_CAN_REWARD_RAGE,
     UNITHOOK_PROVIDE_FINAL_SPELL_DAMAGE,
+    UNITHOOK_IS_SPELL_DIMINISHING_EXEMPT,
     UNITHOOK_END
 };
 
@@ -86,6 +87,8 @@ public:
     virtual bool ProvideFinalSpellDamage(Spell* /*cast*/, Unit* /*attacker*/, Unit* /*victim*/,
         SpellInfo const* /*info*/, uint8 /*effectMask*/, uint32& /*damage*/, bool& /*critical*/) { return false; }
 
+    virtual bool IsSpellDiminishingExempt(Spell* /*cast*/, Unit* /*caster*/, Unit* /*target*/, SpellInfo const* /*info*/) { return false; }
+
     // Called when Heal is Recieved
     virtual void ModifyHealReceived(Unit* /*target*/, Unit* /*healer*/, uint32& /*heal*/, SpellInfo const* /*spellInfo*/) { }
 
--- a/src/server/game/Scripting/ScriptMgr.h
+++ b/src/server/game/Scripting/ScriptMgr.h
@@ -569,6 +569,7 @@ public: /* UnitScript */
     void OnDamage(Unit* attacker, Unit* victim, uint32& damage);
     void ModifyPeriodicDamageAurasTick(Unit* target, Unit* attacker, uint32& damage, SpellInfo const* spellInfo);
     void ModifyMeleeDamage(Unit* target, Unit* attacker, uint32& damage);
+    bool IsSpellDiminishingExempt(Spell* cast, Unit* caster, Unit* target, SpellInfo const* info);
     bool ProvideFinalSpellDamage(Spell* cast, Unit* attacker, Unit* victim, SpellInfo const* info, uint8 effectMask, uint32& damage, bool& critical);
     void ModifySpellDamageTaken(Unit* target, Unit* attacker, int32& damage, SpellInfo const* spellInfo);
     void ModifyHealReceived(Unit* target, Unit* healer, uint32& addHealth, SpellInfo const* spellInfo);
--- a/src/server/game/Spells/Spell.cpp
+++ b/src/server/game/Spells/Spell.cpp
@@ -3208,6 +3208,8 @@ SpellMissInfo Spell::DoSpellHitOnUnit(Unit* unit, uint32 effectMask, bool scaleA
     // Get Data Needed for Diminishing Returns, some effects may have multiple auras, so this must be done on spell hit, not aura add
     // Xinef: Do not increase diminishing level for self cast
     m_diminishGroup = GetDiminishingReturnsGroupForSpell(m_spellInfo, m_triggeredByAuraSpell.spellInfo);
+    if (unitCaster && sScriptMgr->IsSpellDiminishingExempt(this, unitCaster, unit, m_spellInfo))
+        m_diminishGroup = DIMINISHING_NONE;
     // xinef: do not increase diminish level for bosses (eg. Void Reaver silence is never diminished)
     if (((m_spellFlags & SPELL_FLAG_REFLECTED) && !(unit->HasReflectSpellsAura())) || (aura_effmask && m_diminishGroup && unit != m_caster && (!unitCaster || !unitCaster->IsCreature() || !unitCaster->ToCreature()->isWorldBoss())))
     {
